proposals/22-03-value-in-the-structured-data.md

```diff
@@ -0,0 +1,511 @@
+Use the status field to record the results of the latest review:
+- implementable: The default for this repo. If the proposal is merged, you can start working on it.
+- deferred: The proposal may be accepted in the future, but it has been shelved for the time being.
+A new PR must be opened to update the proposal and gain reviewer consensus before work can begin.
+- withdrawn: The author changed their mind and no longer wants to pursue the proposal.
+A new PR must be opened to update the proposal and gain reviewer consensus before work can begin.
+- rejected: This proposal should not be implemented.
+- replaced: If you submit a new proposal that supersedes an older one,
+update the older one's status to "replaced by <link>".
+-->
+
+## Summary
+
+This proposal decides the interfaces to change values in the structured data (like json,yaml) inside a Kubernetes objects' field value and implements changing function target a few formats (mainly json).
+
+
+## Motivation
+
+<!--
+If this proposal is an expansion of an existing GitHub issue, link to it here.
+-->
+
+kustomize can apply structured edits to Kubernetes objects defined in yaml files.\
+Sometimes structured multi-line or long single line string (ex. json,yaml, and other structured format data) is injected in Kubernetes objects' string literal field. From the kustomize perspective, these "structured" multiline strings are just arbitrary unstructured strings.\
+So, kustomize can't manipulate one value on structured, formatted data in the Kubernetes object's string literal field. This function is expected behavior, but kustomize will be very helpful if it can change the value of structured data like json and yaml substrings in a string literal.\
+While kustomize won't support unstructured edits in general, theis proposal allows editing of an unstructured string literal as an exception if the string literal is a structured string of a well-known format like json.\
+
+For example, kustomize can't change the value `"REPLACE_TARGET_HOSTNAME"` in this yaml file straightforwardly.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: target-configmap
+data:
+  config.json: |-
+    {"config": {
+      "id": "42",
+      "hostname": "REPLACE_TARGET_HOSTNAME"
+    }}
+```
+
+This function has been wanted for a long time.
+- https://github.com/kubernetes-sigs/kustomize/issues/680
+- https://github.com/kubernetes-sigs/kustomize/issues/3787
+- https://github.com/kubernetes-sigs/kustomize/issues/4517
+
+This function can be an alternative deprecated [vars](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/vars/) function in many use cases.
+
+**Goals:**
+<!--
+List the specific goals of the proposal. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+1. Provide the way to update values in the structured data like kubernetes objects with many formats.
+1. Be able to add replacement able format after.
+
+
+**Non-goals:**
```

Comment:
Also, we should scope out customizing the field merge strategies for the embedded data, and we should add a brief section to the proposal outlining what that means. For example, lists are always treated as atomic.

Reply:
> customizing the field merge strategies for the embedded data

I'm feeling too sorry, I can't understand this mean clearly.
Could you give me additional info like examples?

Reply:
I'm only guessing, but maybe @KnVerey wants to scope out how the inline strings are merged from this KEP. 

For `behavior: merge` on `configMapGenerator`, this feature will merge the 2 inline, say yaml, strings as kustomize has always merged 2 yaml configurations, instead of allowing the user to specify the merge strategy. In such a case, 

> lists are always treated as atomic

for example.